prov/verbs/src/fi_verbs.c

```diff
@@ -2399,22 +2401,35 @@ static int fi_ibv_cq_reset(struct fid_cq *cq, const void *cond)
 }
 
 static inline int
-fi_ibv_poll_events(struct fi_ibv_cq *_cq, int timeout)
+fi_ibv_poll_events(struct fi_ibv_cq *_cq, int *signal, int timeout)
 {
 	int ret;
 	void *context;
+	struct pollfd fds[2];
 
-	ret = fi_poll_fd(_cq->channel->fd, timeout);
+	fds[0].fd = _cq->channel->fd;
+	fds[1].fd = _cq->signal_fd[0];
+
+	fds[0].events = fds[1].events = POLLIN;
+
+	ret = poll(fds, 2, timeout);
 	if (ret == 0)
 		return -FI_EAGAIN;
 	else if (ret < 0)
-		return ret;
+		return -errno;
 
-	ret = ibv_get_cq_event(_cq->channel, &_cq->cq, &context);
-	if (ret)
-		return ret;
+	if (fds[1].revents & POLLIN) {
+		*signal = 1;
```

Comment:
I think returning EAGAIN is fine, even if we were signaled.  There's a race between signaling the thread to return and it timing out anyway.

But, we still need to read the fd.  Data has been written to the socketpair and has never been read of.